app/controllers/requests.rb

```diff
@@ -0,0 +1,89 @@
+class MakersBnb < Sinatra::Base
+
+  get '/requests' do
+    if current_user
+      @users_requests = Request.all(user_id: current_user.id)
+      @received_requests = Request.all(owner_id: current_user.id)
+      @request_display = @users_requests.map do |x|
+        Space.get(x.space_id)
+      end
+      erb :'requests/index'
+    else
+      redirect to '/sessions/new'
+    end
+  end
+
+  post '/requests/new' do
+
+    requested_date = Date.parse(params[:date_requested]) rescue nil
+
+    unless requested_date
+      flash[:notice] = 'Please enter a valid date'
+      redirect '/spaces'
+    end
+
+    unless session[:available_dates].include?(Date.parse(params[:date_requested]).strftime)
+      flash[:notice] = 'Space not available on the requested date'
+      redirect '/spaces'
+    end
+
+    if current_user
+      unless current_user.id != Space.get(params[:space_id]).user_id
+        flash[:notice] = 'This is your own space'
+        redirect '/spaces'
+      end
+      date = Date.parse(params[:date_requested])
+      space = Space.get(params[:space_id])
+      request = Request.create(user_id: current_user.id, space_id: params[:space_id], date_requested: date, owner_id: space.user_id)
+      flash.keep[:notice] = 'Booking requested'
+      redirect '/users/account'
+    else
+      flash[:notice] = 'Please log in to request a space'
+      redirect '/spaces'
+    end
+  end
```

Comment:
Generally, ok. But this controller action is far too long - your code has gone all procedural instead of tight OO. Really, this controller action should say something like this (see if you can work out what I mean):

```ruby
post '/requests/new' do
  return redirect to '/spaces', notice: 'Please enter a valid date' unless Date.valid_date? params[:date_requested]
  @request = Request.try_create(params[:space_id], current_user, params[:date_requested])
  return redirect to '/users/account', notice: 'Booking requested' if @request.id # if it has an id, it's in the database
  return redirect to '/space', notice: @request.errors # save any errors onto an errors object on the request
end
```